pytorch/main.py

```diff
@@ -212,79 +212,104 @@ def train(train_loader, model, ema_model, optimizer, epoch, log):
-            meters.update('cons_loss', consistency_loss.data[0])
-        else:
-            consistency_loss = 0
-            meters.update('cons_loss', 0)
-
-        loss = class_loss + consistency_loss + res_loss
-        assert not (np.isnan(loss.data[0]) or loss.data[0] > 1e5), 'Loss explosion: {}'.format(loss.data[0])
-        meters.update('loss', loss.data[0])
-
-        prec1, prec5 = accuracy(class_logit.data, target_var.data, topk=(1, 5))
-        meters.update('top1', prec1[0], labeled_minibatch_size)
-        meters.update('error1', 100. - prec1[0], labeled_minibatch_size)
-        meters.update('top5', prec5[0], labeled_minibatch_size)
-        meters.update('error5', 100. - prec5[0], labeled_minibatch_size)
-
-        ema_prec1, ema_prec5 = accuracy(ema_logit.data, target_var.data, topk=(1, 5))
-        meters.update('ema_top1', ema_prec1[0], labeled_minibatch_size)
-        meters.update('ema_error1', 100. - ema_prec1[0], labeled_minibatch_size)
-        meters.update('ema_top5', ema_prec5[0], labeled_minibatch_size)
-        meters.update('ema_error5', 100. - ema_prec5[0], labeled_minibatch_size)
-
-        # compute gradient and do SGD step
         optimizer.zero_grad()
-        loss.backward()
+
+        # For each sub-batch, accumulate gradients. But only do one optimization step
+        # at the end.
+        #
+        # Note that because of `size_average=False` in the call to cross entropy,
+        # then accumulating (rather than averaging) the gradients is correct.
+        for input, ema_input, target in zip(input_chunks, ema_input_chunks, target_chunks):
+            # measure data loading time
+            meters.update('data_time', time.time() - end)
+
+            input_var = torch.autograd.Variable(input)
+            ema_input_var = torch.autograd.Variable(ema_input, volatile=True)
+            target_var = torch.autograd.Variable(target.cuda(**{'async': True}))
+
+            minibatch_size = len(target_var)
+            labeled_minibatch_size = target_var.data.ne(NO_LABEL).sum()
+            assert labeled_minibatch_size > 0
+            meters.update('labeled_minibatch_size', labeled_minibatch_size)
+
+            ema_model_out = ema_model(ema_input_var)
+            model_out = model(input_var)
+
+            if isinstance(model_out, Variable):
+                assert args.logit_distance_cost < 0
+                logit1 = model_out
+                ema_logit = ema_model_out
+            else:
+                assert len(model_out) == 2
+                assert len(ema_model_out) == 2
+                logit1, logit2 = model_out
+                ema_logit, _ = ema_model_out
+
+            ema_logit = Variable(ema_logit.detach().data, requires_grad=False)
+
+            if args.logit_distance_cost >= 0:
+                class_logit, cons_logit = logit1, logit2
+                res_loss = args.logit_distance_cost * residual_logit_criterion(class_logit, cons_logit) / minibatch_size / 2
```

Comment:
You don't need to divide by two manually here, as `minibatch_size` is already half of what it originally is, since it's computed from `len(target_var)`, which is a chunk.

Reply:
Also, there is no `from __future__ ...` here, so make sure to run it with py3.

Reply:
You are right, I was mistaken. Just for reference, here's the pseudo-math how I verified it.

```
grad = 0
original code: grad += sum_100 loss_i / 100  => 1 if all loss_i = 1
two-step: grad += sum_50 loss_i / 50 / 2  ; grad += sum_50 loss_i / 50 / 2  => 0.5 + 0.5 if all loss_i = 1
```